### PR TITLE
Support any number of data products from the command line

### PR DESCRIPTION
**Pull request by duytnguyendtn**
### Description

More quality of life improvements continuing after #1887 in prep of the jdaviz installer effort. This PR does two things:

1. The more important one: Removes the data/filepath argument requirement for the desktop app, effectively allowing the user to launch an "empty" config from the command line (e.g. `jdaviz specviz`)
2. ~Adds direct launchers for each config, allowing the user to run `specviz` instead of `jdaviz specviz`. These two commands are now equivalent~ (This was removed as per discussions: https://github.com/spacetelescope/jdaviz/pull/1960#issuecomment-1444405656)
3. The CLI interface now allows multiple, simultaneous data products to be specified for all configs (whether the config will accept multiple data products depends on the config (e.g. Cubeviz)). This was refactored to avoid the Mosviz loading error found in https://github.com/spacetelescope/jdaviz/pull/1960#pullrequestreview-1320497884
4. VERY IMPORTANT FOR DEVS: Centralizes all `voila` notebooks to a central `jdaviz_cli.ipynb` and REMOVES the config-specific voila notebooks

This is all CLI stuff, so we don't have tests; I expect this to decrease coverage slightly. If anyone has any ideas on how to test this, please let me know!

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/cubeviz/cubeviz.ipynb [hunk 1]
```diff
@@ -10,14 +10,16 @@
...
+    "if data_path:\n",
+    "    cubeviz.load_data('DATA_FILENAME')\n",
     "cubeviz.app"
    ]
   }
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3",
+   "display_name": "Python 3.10.7 ('envmain': venv)",
```
Comment (kecnry) [inline jdaviz/configs/cubeviz/cubeviz.ipynb:22]: not sure if this will cause any problems, but might want to clean all the meta-data changes in the notebooks out of the diff?

File: setup.cfg [hunk 2]
```diff
@@ -73,6 +73,11 @@ jdaviz.configs.imviz.tests = data/*
 [options.entry_points]
 console_scripts =
     jdaviz = jdaviz.cli:_main
+    specviz = jdaviz.cli:_specviz
+    specviz2d = jdaviz.cli:_specviz2d
+    imviz = jdaviz.cli:_imviz
+    cubeviz = jdaviz.cli:_cubeviz
+    mosviz = jdaviz.cli:_mosviz
```
Comment (kecnry) [inline setup.cfg:80]: awesome!  (Although maybe we should omit mosviz for now since we can't actually load anything directly from the UI afaik)

File: jdaviz/cli.py [hunk 3]
```diff
@@ -85,15 +88,17 @@ def main(filename, layout='default', instrument=None, browser='default',
...
     parser = argparse.ArgumentParser(description='Start a Jdaviz application instance with data '
                                      'loaded from FILENAME.')
-    parser.add_argument('layout', choices=['cubeviz', 'specviz', 'specviz2d', 'mosviz', 'imviz'],
-                        help='Configuration to use.')
-    parser.add_argument('filename', type=str,
+    if config is None:
+        parser.add_argument('layout', choices=['cubeviz', 'specviz', 'specviz2d',
+                                               'mosviz', 'imviz'],
+                            help='Configuration to use.')
+    parser.add_argument('--filename', type=str, default=None,
```
Comment (kecnry) [inline jdaviz/cli.py:101]: is this intentionally `--filename` now?  That seems to break passing filename as a positional argument.

**Review by kecnry — COMMENTED** [review]
This will be great to have!

I'm getting an unrecognized argument error when attempting to pass a filename to `jdaviz cubeviz` or `cubeviz` as a positional argument though... I don't know if we want to break support for that.

File: jdaviz/cli.py (hunk 3 above)
Comment (duytnguyendtn) [reply jdaviz/cli.py:101]: I don't think I changed those... This is on main: https://github.com/spacetelescope/jdaviz/blob/4366f754026d2ba3745e70b0a35634a413375668/jdaviz/cli.py#L96

File: setup.cfg (hunk 2 above)
Comment (duytnguyendtn) [reply setup.cfg:80]: Can you specify a directory via the file picker? I actually don't know 😅

File: jdaviz/configs/cubeviz/cubeviz.ipynb (hunk 1 above)
Comment (duytnguyendtn) [reply jdaviz/configs/cubeviz/cubeviz.ipynb:22]: Grr... I wish I could just turn that metadata off my IDE... Thanks for the catch, I'll clean that up

File: jdaviz/cli.py (hunk 3 above)
Comment (duytnguyendtn) [reply jdaviz/cli.py:101]: Oh wait, you're talking about the `--` part. Yes that's required for the change, as that's how `argparse` specifies an optional argument. Why isn't specifying a "default" keyword sufficient? I don't know...

File: jdaviz/cli.py [hunk 4]
```diff
@@ -85,15 +88,17 @@ def main(filename, layout='default', instrument=None, browser='default',
         os.chdir(start_dir)
 
 
-def _main():
+def _main(config=None):
```
Comment (pllim) [inline jdaviz/cli.py:91]: This change seems unrelated. What is the motivation for this?

Comment (duytnguyendtn) [reply jdaviz/cli.py:91]: Each one of the config launchers call `_main` with their `config` specified. Notice how they're calling `_main` and not `main` (no underline) to handle the argument parsing for them

File: jdaviz/cli.py (hunk 3 above)
Comment (pllim) [reply jdaviz/cli.py:101]: I don't think we need this at all. I think we can make the existing filename with nargs set to "no argument is acceptable".

https://docs.python.org/3/library/argparse.html#nargs

File: jdaviz/cli.py [hunk 5]
```diff
@@ -115,6 +120,31 @@ def _main():
...
+        layout = args.layout
+    else:
+        layout = config
+
+    main(filename=args.filename, layout=layout, instrument=args.instrument, browser=args.browser,
          theme=args.theme, verbosity=args.verbosity, history_verbosity=args.history_verbosity,
          hotreload=args.hotreload)
+
+
+def _specviz():
```
Comment (pllim) [inline jdaviz/cli.py:133]: Do we really need this?

File: setup.cfg (hunk 2 above)
Comment (kecnry) [reply setup.cfg:80]: I think its planned, but currently unsupported

Comment (pllim) [reply setup.cfg:80]: AFAIK the current bundled file picker don't take dir as input but I am also not 100% sure. I agree with Kyle that filename should still be mandatory for Mosviz.

File: jdaviz/cli.py (hunk 5 above)
Comment (duytnguyendtn) [reply jdaviz/cli.py:133]: I assume you're referring to the config launchers here? We don't technically need them, but it makes launching the configs much easier from the command line. Until the freeform is more fully fleshed-out, our users are *really* installing `jdaviz` to get access to the configs.

Also, to peel back the curtain a little bit on the installer work, these entry_points create an executable in the virtualenv (AKA the thing you call when you run `$ jdaviz`. I learned you can create desktop shortcuts to these executables so you can "double click" them, and they'll launch. Having separate `specviz.exe`, `cubeviz.exe`, `imviz.exe` executables allows me to simply create a shortcut to the relevant executable. If we don't have these config launchers, I'll need to create shortcuts to the existing `jdaviz.exe` and pass the config as an argument in each shortcut. Doable, but more complex

Comment (kecnry) [reply jdaviz/cli.py:133]: The other options from a desktop shortcut would be to launch a splash page where you then choose which config to load.  I have slight concerns about eventual clashes as we make more configs, but think its worth it for the convenience (and maybe we could eventually have most customized installation where the user picks which to setup/install).

Comment (pllim) [reply jdaviz/cli.py:133]: I am also scared of having so many executables from one install. Also please make sure they go away cleanly on uninstall.

Comment (pllim) [reply jdaviz/cli.py:133]: Sure, but if I uninstall from that environment, I also expect them to die in that environment without having to delete the whole environment to get rid of them. If they are registered properly with Python during install, should just happen but best to double check.

Comment (duytnguyendtn) [reply jdaviz/cli.py:133]: All gone!

```
(envtest) PS E:\STScI\gitRepos\jdaviz> pip uninstall jdaviz
Found existing installation: jdaviz 3.2.dev175+ge3b9d944.d20221205
Uninstalling jdaviz-3.2.dev175+ge3b9d944.d20221205:
  Would remove:
    e:\stsci\gitrepos\jdaviz\envtest\lib\site-packages\jdaviz-3.2.dev175+ge3b9d944.d20221205.dist-info\*
    e:\stsci\gitrepos\jdaviz\envtest\lib\site-packages\jdaviz\*
    e:\stsci\gitrepos\jdaviz\envtest\scripts\cubeviz.exe
    e:\stsci\gitrepos\jdaviz\envtest\scripts\imviz.exe
    e:\stsci\gitrepos\jdaviz\envtest\scripts\jdaviz.exe
    e:\stsci\gitrepos\jdaviz\envtest\scripts\mosviz.exe
    e:\stsci\gitrepos\jdaviz\envtest\scripts\specviz.exe
    e:\stsci\gitrepos\jdaviz\envtest\scripts\specviz2d.exe
    e:\stsci\gitrepos\jdaviz\envtest\share\jupyter\nbconvert\templates\jdaviz-default\ansi.js
    e:\stsci\gitrepos\jdaviz\envtest\share\jupyter\nbconvert\templates\jdaviz-default\app.html
    e:\stsci\gitrepos\jdaviz\envtest\share\jupyter\nbconvert\templates\jdaviz-default\conf.json
    e:\stsci\gitrepos\jdaviz\envtest\share\jupyter\nbconvert\templates\jdaviz-default\index.html.j2
    e:\stsci\gitrepos\jdaviz\envtest\share\jupyter\nbconvert\templates\jdaviz-default\util.js
    e:\stsci\gitrepos\jdaviz\envtest\share\jupyter\voila\templates\jdaviz-default\index.html.j2
Proceed (Y/n)? y
  Successfully uninstalled jdaviz-3.2.dev175+ge3b9d944.d20221205
(envtest) PS E:\STScI\gitRepos\jdaviz> ls .\envtest\Scripts\*viz.exe
(envtest) PS E:\STScI\gitRepos\jdaviz>
```

File: jdaviz/cli.py (hunk 3 above)
Comment (duytnguyendtn) [reply jdaviz/cli.py]: Great find! Implemented in https://github.com/spacetelescope/jdaviz/pull/1890/commits/9fb79d0e2d7aa8542e1577a68f11b6ad91fb8105. For reference, i tried providing multiple filepaths expecting it to explode, but it actually returns a nice error message:

```
(envmain) PS E:\STScI\gitRepos\jdaviz> specviz ..\..\datasets\spec-0751-52251-0160.fits ..\..\datasets\spec-0751-52251-0160.fits
usage: specviz [-h] [--instrument INSTRUMENT] [--browser BROWSER] [--theme {light,dark}] [--verbosity {debug,info,warning,error}] [--history-verbosity {debug,info,warning,error}] [--hotreload | --no-hotreload]
               [--version]
               [filename]
specviz: error: unrecognized arguments: ..\..\datasets\spec-0751-52251-0160.fits
```

Comment (duytnguyendtn) [reply jdaviz/cli.py]: @kecnry this should preserve the backwards compatibility and not require the `--filename` flag

File: setup.cfg (hunk 2 above)
Comment (duytnguyendtn) [reply setup.cfg:80]: Put in the filepath requirement for mosviz in https://github.com/spacetelescope/jdaviz/pull/1890/commits/e3a6d4b82ea23340b1b353aa862f5980ee095868

File: jdaviz/cli.py (hunk 5 above)
Comment (duytnguyendtn) [reply jdaviz/cli.py]: As per the plan in https://github.com/spacetelescope/jdaviz/pull/1960#issuecomment-1444405656, I reverted this change, so this point is now not applicable

File: jdaviz/jdaviz_cli.ipynb [hunk 6]
```diff
@@ -7,17 +7,18 @@
    "outputs": [],
    "source": [
     "# PREFIX\n",
-    "from jdaviz import Cubeviz\n",
+    "from jdaviz import CONFIG\n",
     "\n",
-    "cubeviz = Cubeviz(verbosity='JDAVIZ_VERBOSITY', history_verbosity='JDAVIZ_HISTORY_VERBOSITY')\n",
-    "cubeviz.load_data('DATA_FILENAME')\n",
-    "cubeviz.app"
+    "viz = CONFIG(verbosity='JDAVIZ_VERBOSITY', history_verbosity='JDAVIZ_HISTORY_VERBOSITY')\n",
```
[the remark this answers is not in the document]
Comment (duytnguyendtn) [reply jdaviz/jdaviz_cli.ipynb:12]: It does! https://github.com/spacetelescope/jdaviz/blob/ae518dcdb10530251a3f909d8c1790b72029b45b/jdaviz/cli.py#L78

**Review by rosteen — APPROVED** [review]
I'm very happy to have the notebooks consolidated. I was able to load all the configs (I tried Mosviz with NIRISS data but not with NIRSpec), code looks like code. Approved!

File: CHANGES.rst [hunk 7]
```diff
@@ -4,6 +4,11 @@
 New Features
 ------------
 
+- CLI launchers no longer require data to be specified [#1890]
```
Comment (pllim) [inline CHANGES.rst:7]:
```suggestion
- CLI launchers no longer require data to be specified. [#1890]
```

File: CHANGES.rst [hunk 8]
```diff
@@ -4,6 +4,11 @@
...
+
+- Configurations that support multiple, simultaneous data files now allow
+  multiple data products to be specified in the command line [#1890]
```
Comment (pllim) [inline CHANGES.rst:10]:
```suggestion
  multiple data products to be specified in the command line. [#1890]
```

File: docs/imviz/import_data.rst [hunk 9]
```diff
@@ -19,12 +19,13 @@ See :meth:`jdaviz.configs.imviz.helper.Imviz.load_data` for more information.
 Importing data through the Command Line
 =======================================
 
-When running the Imviz application via the command line, you must provide a path
-to a compatible file, which will be loaded into the app on initialization:
+When running the Imviz application via the command line, you may provide a path
+to a compatible file, which will be loaded into the app on initialization.
+Multiple data files can be provided
```
Comment (pllim) [inline docs/imviz/import_data.rst:24]:
```suggestion
Multiple data files may be provided:
```

File: docs/mosviz/import_data.rst [hunk 10]
```diff
@@ -67,6 +67,8 @@ and for NIRISS:
 
     jdaviz mosviz /path/to/my/data --instrument=niriss
 
+When loading data from the command line, specifying a data file is optional.
```
Comment (pllim) [inline docs/mosviz/import_data.rst:70]: Do you mean directory?

File: jdaviz/cli.py [hunk 11]
```diff
@@ -48,11 +48,17 @@ def main(filename, layout='default', instrument=None, browser='default',
...
-    # Support comma-separate file list
-    filepath = ','.join([str(pathlib.Path(f).absolute()).replace('\\', '/')
-                         for f in filename.split(',')])
+    if filepaths:
+        # Convert paths to posix string; windows paths are not JSON compliant
+        file_list = [pathlib.Path(f).absolute().as_posix() for f in filepaths]
+        if layout == "imviz":
+            # Imviz multiloading should be done all at once for batch processing.
+            # Imviz convention is a single string of consecutive, comma-separated file paths
+            file_list = [','.join(file_list)]
```
Comment (pllim) [inline jdaviz/cli.py:57]: If you really don't like this, you can also modify the Imviz parser itself, though this is probably easier. So 👍 from me.

File: jdaviz/cli.py [hunk 12]
```diff
@@ -48,11 +48,17 @@ def main(filename, layout='default', instrument=None, browser='default',
...
+    else:
+        file_list = []
 
-    with open(os.path.join(CONFIGS_DIR, layout, layout + '.ipynb')) as f:
+    with open(str(JDAVIZ_DIR / "jdaviz_cli.ipynb")) as f:
```
Comment (pllim) [inline jdaviz/cli.py:61]: `open` takes `Path` object, so casting to `str` is unnecessary.

```suggestion
    with open(JDAVIZ_DIR / "jdaviz_cli.ipynb") as f:
```

**Review by pllim — COMMENTED** [review]
I like PRs that remove more lines than they add! 😸 

Just some minor comments.

File: docs/mosviz/import_data.rst (hunk 10 above)
Comment (duytnguyendtn) [reply docs/mosviz/import_data.rst:70]: Oh yeah I guess we can only load directories from the cli right now. May as well change that over. Let me fix that

Comment (duytnguyendtn) [reply docs/mosviz/import_data.rst:70]: Hold up, I think this is actually wrong. I think we actually do require a directory for cli loading... let me check

File: jdaviz/cli.py (hunk 11 above)
Comment (duytnguyendtn) [reply jdaviz/cli.py:57]: I think we should revisit this in a separate push to incorporate batch processing across jdaviz. I know @kecnry suggested just using `jdaviz.batch_load` now, but I'm cautious to do that now (For ex. what if someone provides multiple mosviz directories? Does the `instrument` keyword get passed properly?). I'd prefer to keep the PRs smaller and scope-focused

File: docs/mosviz/import_data.rst (hunk 10 above)
Comment (duytnguyendtn) [reply docs/mosviz/import_data.rst]: Mosviz indeed requires data from the cli. Verified and fixed. Thanks for the catch!

**Review by pllim — APPROVED** [review]
LGTM. Thanks! 

Feel free to merge when CI passes. Ignore dev failures.

**Comment by pllim** [comment]
It should now fail with ASDF if you rebase properly. 😆

**Comment by pllim** [comment]
p.s. Please do "Squash and Merge"

**Comment by duytnguyendtn** [comment]
Test failure in 3.10 seems to be an SDSS query issue; unrelated to this PR. Polled team with no objections to merge.

Thanks Ricky, Pey-Lian, and Kyle for the eyes!
